docs/replication_modes.rst

```diff
@@ -79,6 +93,24 @@ Patroni will only assign one or more synchronous standby nodes based on ``synchr
 On each HA loop iteration Patroni re-evaluates synchronous standby nodes choice. If the current list of synchronous standby nodes are connected and has not requested its synchronous status to be removed it remains picked. Otherwise the cluster member available for sync that is furthest ahead in replication is picked.
 
 
+.. _quorum_mode:
+
+Quorum commit mode
+==================
+
+Starting from PostgreSQL v10 Patroni supports quorum-based synchronous replication.
+
+In this mode, Patroni maintains synchronization state in the DCS, containing the latest known primary, the number of nodes required for quorum and the nodes currently eligible to vote on quorum. In steady state, the nodes voting on quorum are the leader and all synchronous standbys. This state is updated with strict ordering constraints, with regards to node promotion and ``synchronous_standby_names``, to ensure that at all times any subset of voters that can achieve quorum includes at least one node with the latest successful commit.
+
+On each iteration of HA loop, Patroni re-evaluates synchronous standby choices and quorum, based on node availability and requested cluster configuration. In PostgreSQL versions above 9.6 all eligible nodes are added as synchronous standbys as soon as their replication catches up to leader.
+
+Quorum commit helps to reduce worst case latencies, even during normal operation, as a higher latency of replicating to one standby can be compensated by other standbys.
```

Comment:
> My understanding so far is that roughly:
`synchronous_mode: on`: sets `synchronous_standby_names`: `<synchronous_node_count> (<list_of_standbys>)`

Almost. It will be not a `<synchronous_node_count>`, but `len(<list_of_standbys>)`, where `<list_of_standbys>` are nodes that are eligible (state == 'running', don't have `nosync` tag, and don't lag more than `maximum_lag_on_syncnode`) to be synchronous and the total number of nodes in the list will not exceed `synchronous_node_count`.

> `synchronous_mode: quorum`: sets `synchronous_standby_names:` `ANY <synchronous_node_count> (<list_of_ALL_standbys>)`

Also almost.
- `<list_of_ALL_standbys>` -- not all nodes, but nodes with state == 'running', without `nosync` tag, and not lagging more than `maximum_lag_on_syncnode`.
- `synchronous_node_count` -- there will be a value `N`: `1 <= N <= min(synchronous_node_count, len(<list_of_standbys>))`

The main difference between `synchronous_mode` `on` and `quorum` that in the first case Patroni picks a fixed number of synchronous nodes and they are all synchronous. During failover with `synchronous_mode: on` we can **safely promote any node** from this  list and there will be no client visible data loss.

With `quorum` there will be something like `ANY 2 (n1, n2, n3)` in `synchronous_standby_names` and in order to guaranty that there is no data loss during failover Patroni has to check some minimum subset of nodes from this list to prove that the given node received not less WAL stream than others.
Example:
primary: n0
quorum: ANY 2 (n1, n2, n3)

That is, two nodes from this least could be promoted without client visible data loss, and every node need to figure out whether it has all data.
Lets say n0 and n1 are failed. But, we know that at least two nodes from n1, n2, n3 have the latest data, therefore at least one from n2, n3 also has it. Therefore n2 **must** see n3 and compare LSNs to make decision about promotion. If n2 is not behind of n3, than n2 could be promoted and there will be no client visible data loss.

Reply:
Cool, thanks for clarifying!

I guess it would be valuable to describe this in the docs, even if with a simpler approach, just so users can getter a better understanding of the difference between each of the synchronous modes.

BTW I intend to get back to reviewing this PR today.